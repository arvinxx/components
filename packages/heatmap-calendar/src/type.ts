/**
 * 热力图数据结构
 */
export interface ItemData {
  /**
   * 日期
   */
  date: string;
  /**
   * 天数
   */
  day: number;
  /**
   * 月份
   */
  month: number;
  /**
   * 热点数
   */
  total: number;
  /**
   * 周
   */
  week: number;
  /**
   * 年
   */
  year: number;
}

export type HeatmapCalendarData = ItemData[];
