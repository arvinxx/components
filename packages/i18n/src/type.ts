import type { ComponentProps, FC } from 'react';
import type { IntlProvider } from 'react-intl';

export type LocalesType = 'en-US' | 'zh-CN';

export type LocaleKey = string;

export type IntlProviderProps<T = Record<string, string>> = FC<
  Omit<ComponentProps<typeof IntlProvider>, 'messages'> & {
    messages: T;
  }
>;
