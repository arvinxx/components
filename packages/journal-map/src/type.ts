/**
 * 阶段
 */
export interface Step<T = string> {
  id: T;
  name: string;
  /**
   * 该步骤的颜色
   */
  color?: string;
}

/**
 * 用户旅程地图数据
 */
export interface JournalMapData<T extends string = string> {
  /**
   * 大的步骤
   */
  steps: Step<T>[];
  /**
   * 用户行为
   */
  actions: Record<T, UserAction[]>;
}

/**
 * 用户行为
 */
export interface UserAction {
  /**
   * 行为名称
   */
  title: string;
  /**
   * 说明
   */
  description?: string;

  /**
   * 有关的想法
   */
  thoughts?: string[];
  /**
   * 情绪得分
   */
  emotion: number;
  /**
   * 痛点
   */
  painSpot?: string[];
}
