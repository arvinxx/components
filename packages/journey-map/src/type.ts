/**
 * 阶段
 */
export interface Step<T = string> {
  /**
   * 阶段id
   * @TJS-type string
   */
  id: T;
  /**
   * 阶段名称
   */
  name: string;
  /**
   * 该阶段的颜色
   */
  color?: string;
}

/**
 * 用户旅程地图数据
 */
export interface JourneyMapData<T extends string = string> {
  /**
   * 大的步骤
   */
  steps: Step<T>[];
  /**
   * 用户行为
   */
  actions: Record<T, UserAction[]>;
}

/**
 * 用户行为
 */
export interface UserAction {
  /**
   * 行为名称
   */
  title: string;
  /**
   * 说明
   */
  description?: string;

  /**
   * 有关的想法
   */
  thoughts?: string[];
  /**
   * 情绪得分
   */
  emotion: number;
  /**
   * 痛点
   */
  painSpot?: string[];
}

/**
 * 解析后的 YML 数据
 */
export interface JourneyMapYML<T extends string = string> {
  /**
   * 步骤
   */
  steps: Step<T>[];
  /**
   * 用户行为
   */
  actions: Record<T, Record<T, number>[]>;
}
