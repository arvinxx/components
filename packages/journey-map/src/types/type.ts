/**
 * 阶段
 */
export interface Stage<T = string> {
  /**
   * 阶段id
   * @TJS-type string
   */
  id: T;
  /**
   * 阶段名称
   */
  name: string;
  /**
   * 该阶段的颜色
   */
  color?: string;
}

/**
 * 用户旅程地图数据
 */
export interface JourneyMapData<T extends string = string> {
  /**
   * 大的步骤
   */
  stages: Stage<T>[];
  /**
   * 用户行为
   */
  actions: Record<T, UserAction[]>;
}

/**
 * 用户行为类型
 */
export interface UserAction {
  /**
   * 行为名称
   */
  name: string;
  /**
   * 情绪得分
   * @minimum -2
   * @maximum 2
   * @TJS-type integer
   */
  emotion?: number;

  /**
   * 说明
   */
  description?: string;

  /**
   * 想法
   */
  thoughts?: string[];

  /**
   * 痛点
   */
  painSpot?: string[];
}
