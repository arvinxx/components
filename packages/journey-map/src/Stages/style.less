.avx-journey-map-stage {
  &-container {
    display: flex;
    flex: 1;
  }

  &-content {
    display: flex;
    flex: 1;
  }

  &-step {
    display: flex;
    align-items: center;
    justify-content: center;
    padding: 8px;
    border-radius: 4px;

    &-text {
      height: 32px;
      line-height: 32px;
      text-align: center;
    }
  }
}
