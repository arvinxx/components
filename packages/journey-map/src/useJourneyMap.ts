import useMergeValue from 'use-merge-value';
import { useEffect, useState } from 'react';

import { getServiceToken, YMLToJSON } from './utils';
import type { YMLJourneyMapData, JourneyMapData } from './types';

interface StoreParams {
  data: JourneyMapData | string;
  onChange?: (data: JourneyMapData) => void;
  title?: string;
}

export const useJourneyMap = ({
  data,
  onChange,
  title: initTitle,
}: StoreParams) => {
  const defaultJourneyMap: JourneyMapData = { stages: [], actions: {} };

  const hook = useMergeValue<JourneyMapData>(
    typeof data === 'string' ? defaultJourneyMap : data,
    { onChange },
  );

  const [title, setTitle] = useState<string>(initTitle);

  const [store, setStore] = hook;
  /**
   * 从 YML URL 中解析 JSON
   * @param url
   */
  const fetchYMLToJSON = async (url: string): Promise<YMLJourneyMapData> => {
    const res = await fetch(url);
    const yml = await res.text();

    return YMLToJSON(yml);
  };

  useEffect(() => {
    if (typeof data === 'string' && data.startsWith('http')) {
      fetchYMLToJSON(data).then((result) => {
        const { stages, actions } = result;
        setTitle(result.title);
        setStore({ stages, actions });
      });
    }
  }, [data]);

  const actionList = Object.values(store.actions).flat();

  return { store, actionList, actionLength: actionList.length, title };
};

export const JourneyMapStore = getServiceToken(useJourneyMap);
