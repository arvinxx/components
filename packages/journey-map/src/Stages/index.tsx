import type { FC } from 'react';
import React, { useContext, useRef } from 'react';
import { blue } from '@ant-design/colors';
import { useSize } from 'ahooks';

import { OverflowTitle, Section } from '../components';
import { JourneyMapStore } from '../useJourneyMap';
import { calcStageLength } from '../utils';

import './style.less';

const Stage: FC<{ name: string; color: string }> = ({ name, color }) => {
  const ref = useRef();
  const { width } = useSize(ref);
  const { config } = useContext(JourneyMapStore);

  const height = config?.height?.stage;

  return (
    <div
      ref={ref}
      className="avx-journey-map-stage-step"
      style={{ background: color, height }}
    >
      <div
        className="avx-journey-map-stage-step-text"
        style={{ width: width ? width - 16 : 28 }}
      >
        <OverflowTitle direction={'x'} title={name} />
      </div>
    </div>
  );
};

const Stages: FC = () => {
  const { store, actionLength, config } = useContext(JourneyMapStore);
  const { stages, actions } = store;
  const height = config?.height?.stage || 48;

  return (
    <div className="avx-journey-map-stage-container" style={{ height }}>
      <Section height={height}>阶段</Section>
      <div className="avx-journey-map-stage-content">
        {stages.map((stage, index) => {
          const color = stage.color || blue[0];

          const actionCount = actions[stage.id]?.length || 0;

          const { width, margin } = calcStageLength({
            actionCount,
            stageCount: stages.length,
            index,
            totalCount: actionLength,
          });

          return (
            <div key={stage.id} style={{ width: `${width}%`, ...margin }}>
              <Stage name={stage.name} color={color} />
            </div>
          );
        })}
      </div>
    </div>
  );
};

export default Stages;
