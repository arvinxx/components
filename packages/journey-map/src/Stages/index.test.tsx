import React from 'react';
import { render } from '@testing-library/react';

import Stages from './index';
import { JourneyMapStore, useJourneyMap } from '../useJourneyMap';

import type { FC } from 'react';
import type { Stage } from '../types';

const Demo: FC<{ stages: Stage[] }> = ({ stages }) => {
  const value = useJourneyMap({
    data: {
      actions: {},
      stages,
    },
  });
  return (
    <JourneyMapStore.Provider value={value}>
      <Stages />
    </JourneyMapStore.Provider>
  );
};

describe('Actions', () => {
  it('默认状态', () => {
    const { container } = render(
      <Demo
        stages={[
          { id: '1', name: '1' },
          { id: '2', name: '2', color: 'blue' },
        ]}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
