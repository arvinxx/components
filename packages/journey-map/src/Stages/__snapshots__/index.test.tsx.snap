// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Actions 默认状态 1`] = `
<div>
  <div
    class="avx-journey-map-stage-container"
    style="height: 48px;"
  >
    <div
      class="avx-journey-map-section"
      style="height: 48px;"
    >
      <div
        class="avx-journey-map-section-text"
      >
        阶段
      </div>
    </div>
    <div
      class="avx-journey-map-stage-content"
    >
      <div
        style="width: 50%; margin: 0px 0.5%;"
      >
        <div
          class="avx-journey-map-stage-step"
          style="background: rgb(230, 247, 255);"
        >
          <div
            class="avx-journey-map-stage-step-text"
            style="width: 28px;"
          >
            <div
              style="overflow: hidden; white-space: nowrap; text-overflow: ellipsis;"
            >
              1
            </div>
          </div>
        </div>
      </div>
      <div
        style="width: 50%; margin: 0px 0.5%;"
      >
        <div
          class="avx-journey-map-stage-step"
          style="background: blue;"
        >
          <div
            class="avx-journey-map-stage-step-text"
            style="width: 28px;"
          >
            <div
              style="overflow: hidden; white-space: nowrap; text-overflow: ellipsis;"
            >
              2
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
